src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -390,52 +499,117 @@ internal DkmEvaluationResult GetResult(EvalResultDataItem dataItem, DkmClrType r
+                display = value.GetValueString(inspectionContext, Formatter.NoFormatSpecifiers);
             }
 
-            var typeName = debuggerDisplayType ?? GetTypeName(value.InspectionContext, declaredType, runtimeType);
-            return CreateEvaluationResult(value, name, typeName, display, dataItem);
+            var typeName = displayType ?? GetTypeName(inspectionContext, value, declaredType, dataItem.Kind);
+
+            return CreateEvaluationResult(inspectionContext, value, name, typeName, display, dataItem);
         }
 
-        private DkmGetChildrenAsyncResult GetChildren(DkmInspectionContext inspectionContext, DkmEvaluationResult evaluationResult, EvalResultDataItem dataItem, int initialRequestSize)
+        private void GetChildrenAndContinue(EvalResultDataItem dataItem, DkmWorkList workList, DkmStackWalkFrame stackFrame, int initialRequestSize, DkmInspectionContext inspectionContext, DkmCompletionRoutine<DkmGetChildrenAsyncResult> completionRoutine)
         {
             var expansion = dataItem.Expansion;
-            var builder = ArrayBuilder<DkmEvaluationResult>.GetInstance();
+            var rows = ArrayBuilder<EvalResultDataItem>.GetInstance();
             int index = 0;
             if (expansion != null)
             {
-                expansion.GetRows(this, builder, inspectionContext, dataItem, dataItem.Value, 0, initialRequestSize, visitAll: true, index: ref index);
+                expansion.GetRows(this, rows, inspectionContext, dataItem, dataItem.Value, 0, initialRequestSize, visitAll: true, index: ref index);
             }
-            var rows = builder.ToArrayAndFree();
-            Debug.Assert(index >= rows.Length);
-            Debug.Assert(initialRequestSize >= rows.Length);
-            var enumContext = DkmEvaluationResultEnumContext.Create(index, evaluationResult.StackFrame, evaluationResult.InspectionContext, new EnumContextDataItem(dataItem));
-            return new DkmGetChildrenAsyncResult(rows, enumContext);
+            var numRows = rows.Count;
+            Debug.Assert(index >= numRows);
+            Debug.Assert(initialRequestSize >= numRows);
+            var initialChildren = new DkmEvaluationResult[numRows];
+            GetEvaluationResultsAndContinue(rows, initialChildren, 0, numRows, workList, inspectionContext, stackFrame,
+                () => ContinueWithExceptionHandling(
+                    () =>
+                    {
+                        rows.Free();
+                        var enumContext = DkmEvaluationResultEnumContext.Create(index, stackFrame, inspectionContext, new EnumContextDataItem(dataItem));
+                        completionRoutine(new DkmGetChildrenAsyncResult(initialChildren, enumContext));
+                    },
+                    e =>
+                    {
+                        rows.Free();
+                        completionRoutine(DkmGetChildrenAsyncResult.CreateErrorResult(e));
+                    }));
```

Comment:
If the onCompletion argument to ContinueWithExceptionHandling throws an exception, rows will be freed twice. Same comment for GetItemsAndContinue.